engine/reformulation/core/src/main/java/it/unibz/inf/ontop/answering/reformulation/input/translation/impl/RDF4JInputQueryTranslatorImpl.java

```diff
@@ -405,6 +416,46 @@ private TranslationResult translateOrder(Order node, ImmutableMap<Variable, Grou
         ));
     }
 
+    private TranslationResult translateService(Service service, ImmutableMap<Variable, GroundTerm> externalBindings) throws OntopInvalidInputQueryException {
+        
+        final Set<String> serviceVars = service.getServiceVars();
+        ImmutableSet<Variable> allVars = serviceVars.stream()
+                .map(termFactory::getVariable)
+                .collect(ImmutableCollectors.toSet());
+        String collect = allVars.stream().map(Variable::getName).collect(Collectors.joining(","));
+        SPARQLRepository srq = new SPARQLRepository(service.getServiceRef().getValue().stringValue());
+        
+        Map<String, MapBindingSet> sbs = new LinkedHashMap<>();
+        
+        for (String sv:serviceVars) {
+            final MapBindingSet bindingSet = new MapBindingSet();
+            sbs.put(sv, bindingSet);
+        }
+            
+        srq.init();
+        try (RepositoryConnection conn = srq.getConnection()) {
+            String remoteQuery = service.getSelectQueryString(serviceVars);
+            TupleQuery prepareQuery = conn.prepareTupleQuery(QueryLanguage.SPARQL, remoteQuery);
+            try (TupleQueryResult evaluate = prepareQuery.evaluate()) {
+                while (evaluate.hasNext()) {
+                    BindingSet next = evaluate.next();
+                    service.getServiceVars().forEach(bindingName -> {
+                        sbs.get(bindingName).addBinding(next.getBinding(bindingName));
+                    });
+                }
+            }
+        }
+        Set<BindingSet> bs = new LinkedHashSet<>();
```

Comment:
Shouldn't it be a List instead of a Set?
In principle, the SPARQL subquery could return duplicates.

Reply:
Good question. I always assumed it should be a set because the [sparql-11-federation](https://www.w3.org/TR/sparql11-federated-query/#algebra_service) mentions "the multiset of solution mappings corresponding to the results of executing query". 

Let me confirm that for you.

Reply:
Multiset is another name for bag, it accepts duplicates (unlike sets)